### PR TITLE
[FIX] sale: avoid access error on `credit_limit` field

### PR DESCRIPTION
**Pull request by lvsz**
Versions
--------
- 16.0+

Steps (18.0+)
-------------
1. Enable credit limits via Accounting settings;
2. log in as a user with access to Sales but not Accounting;
3. create a Sales Order;
4. add a product.

Issue
-----
Access Error.

Cause
-----
Commit de302c2 changed the way company dependent fields are handled. Instead of computing them via `_compute_company_dependent`, they are now stored in the database.

Before this this change, any `groups` restriction added to a field wasn't actually checked. After this change, it does get checked, leading to the access error.

Solution
--------
1. Use `sudo` to access `partner_id.credit`.
2. When calling the `_build_credit_warning_message`, pass the sales order with `sudo`.

opw-4367393

### CODE REVIEW

**Comment by Feyensv** [comment]
Honestly kinda surprised there are no tests covering the creation of a SO with only sales right (or maybe those tests exists but without a test Form ?).  Would you mind adding one then ? :pray:

File: addons/sale/models/sale_order.py [hunk 1]
```diff
@@ -589,9 +589,11 @@ def _compute_partner_credit_warning(self):
             show_warning = order.state in ('draft', 'sent') and \
                            order.company_id.account_use_credit_limit
             if show_warning:
-                updated_credit = order.partner_id.commercial_partner_id.credit + (order.amount_total / order.currency_rate)
+                order_sudo = order.sudo()
                 order.partner_credit_warning = self.env['account.move']._build_credit_warning_message(
-                    order, updated_credit)
+                    order_sudo,
+                    order_sudo.partner_id.credit + order.amount_total / order.currency_rate,
```
Comment (Feyensv) [inline addons/sale/models/sale_order.py:595]:
```suggestion
                    order_sudo.partner_id.commercial_partner_id.credit + order.amount_total / order.currency_rate,
```
why did you remove the `.commercial_partner_id` ?

Comment (lvsz) [reply addons/sale/models/sale_order.py]: good question, no idea, re-added it

**Comment by lvsz** [comment]
> Honestly kinda surprised there are no tests covering the creation of a SO with only sales right (or maybe those tests exists but without a test Form ?).

I reckon they exist, just not with `company.account_use_credit_limit` being set to `True`.

> Would you mind adding one then ? 🙏

Done :+1:

File: addons/sale/tests/test_credit_limit.py [hunk 2]
```diff
@@ -63,3 +75,26 @@ def test_credit_limit_multicurrency(self):
...
+
+    @users('notaccountman')
+    def test_credit_limit_access(self):
+        """Ensure credit warning gets displayed without Accounting access."""
+        for group in self.partner_a._fields['credit'].groups.split(','):
+            self.assertFalse(self.env.user.has_group(group))
+
+        self.partner_a.sudo().credit_limit = self.product_a.list_price
+
+        with Form(self.empty_order.with_env(self.env)) as order_form:
```
Comment (Feyensv) [inline addons/sale/tests/test_credit_limit.py:87]: The test didn't fail even when removing the `sudo` on `order_sudo`, because `@users` only changes `self.env`, not `self.empty_order.env`.  Pushed a little fixup commit to fix that ;)

**Review by Feyensv — APPROVED** [review]
@robodoo delegate+

File: addons/sale/tests/test_credit_limit.py (hunk 2 above)
Comment (lvsz) [reply addons/sale/tests/test_credit_limit.py:87]: Good catch, kinda assumed it wouldn't fail yet anyway till the 18.0 forward port.

**Comment by lvsz** [comment]
robodoo r+

File: addons/sale/tests/test_credit_limit.py (hunk 2 above)
Comment (lvsz) [reply addons/sale/tests/test_credit_limit.py:88]: Actually, the test failed in that case because I forgot to set `sale.order.user_id` to the current user, leading to the access error.

Accessing `credit` or `credit_limit` without accounting privileges doesn't result in an access error in 16.0 (yet).

**Comment by lvsz** [comment]
robodoo r+

File: addons/sale/tests/test_credit_limit.py (hunk 2 above)
Comment (Feyensv) [reply addons/sale/tests/test_credit_limit.py:88]: Oh yeah, didn't investigate much, but wanted to make sure the test was covering what we wanted correctly :D
The explanation makes sense, thanks :+1:
